File: leetcode/merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**************************************
* @author dujiong
* @date 2016.9.20
* @version V0.1
**************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        int n=lists.size();
        if(n==0){
            return NULL;
        }
        ListNode *res = lists[0];
        for(int i=1;i<n;i++){
            res=mergeTwoLists(res,lists[i]);
        }
        return res;
    }
    
private:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        if(l1==NULL || l2==NULL){
            return (l1==NULL)?l2 : l1;
        }
        ListNode dummy(0);
        ListNode* p = &dummy;
        while(l1!=NULL && l2!=NULL){
            if(l1->val <= l2->val){
                p->next = l1;
                l1 = l1->next;
            }else{
                p->next = l2;
                l2 = l2->next;
            }
            p = p->next;
        }
        p->next=((l1==NULL)?l2:l1);
        return dummy.next;
    }
};

File: leetcode-golang/Merge k Sorted Lists/Merge k Sorted Lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 //从合并两个链表引申过来的方法
 func mergeKLists(lists []*ListNode) *ListNode {
    if len(lists) == 0 {
        return nil
    }
    if len(lists)
    for i := 1; i < len(lists); i++ {
        lists[i] = mergeTwoLists(lists[i-1], lists[i])
    }
    return lists[len(lists)-1]
}

func mergeTwoLists(l1, l2 *ListNode) *ListNode {
    if l1 == nil && l2 == nil {
        return nil
    }
    dummy := &ListNode{}
    head := dummy
    for l1 != nil && l2 != nil {
        if l1.Val < l2.Val {
            head.Next = l1
            l1 = l1.Next
        } else {
            head.Next = l2
            l2 = l2.Next
        }
        head = head.Next
    }
    if l1 != nil {
        head.Next = l1
    }
    if l2 != nil {
        head.Next = l2
    }
    return dummy.Next
}


